Lib/dataclasses.py

```diff
@@ -1059,12 +1065,36 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
         # Does this class have a post-init function?
         has_post_init = hasattr(cls, _POST_INIT_NAME)
 
+        # Does it only accept kwargs?
+        kw_only_post_init = False
+        if has_post_init:
+
+            # get params
+            post_init_sig = inspect.signature(getattr(cls, _POST_INIT_NAME))
+            post_init_params = post_init_sig.parameters
+
+            # are there more than 1 POSITIONAL_OR_KEYWORD params?
+            # ('self' is convention, not guaranteed spelling)
+            p_or_k_count = 0
+            for param in post_init_params.values():
+                if param.kind == param.POSITIONAL_OR_KEYWORD:
+                    p_or_k_count += 1
+                    if p_or_k_count > 1:
+                        break
+
+            # set to True if __post_init__ has a '*' after self, forcing
+            # all additional params to KEYWORD_ONLY
+            # eg, def __post_init__(self, *, init_var1, init_var2...
+            if p_or_k_count == 1 and len(post_init_params) > 1:
+                kw_only_post_init = True
+
         _set_new_attribute(cls, '__init__',
                            _init_fn(all_init_fields,
                                     std_init_fields,
                                     kw_only_init_fields,
                                     frozen,
                                     has_post_init,
+                                    kw_only_post_init,
```

Comment:
If passing another arg to _init_fn is undesirable, perhaps `has_post_init` could be changed to post_init_type, 0=None, 1=default behavior, 2=kw_only_post_init

Reply:
I just pushed a revised version that re-uses KW_ONLY and passes it via has_post_init